### ISSUE TITLE
Why do we need to respecify barcode information in sc_exon_mapping?

### ISSUE THREAD

**Issue by LTLA** [LuyiTian/scPipe#88]
In `sc_trim_barcode`, the user has to specify the barcode/UMI structure. That's fine.

However, the same information has to be re-specified in `sc_exon_mapping`! This was a major surprise, I had assumed that `sc_exon_mapping` would figure out the barcode and UMI length from the information encoded in the read name. It doesn't help that both of these parameters default to arbitrary values, which results in the function silently continuing on rather than throwing an error when they're not supplied.

It seems as if it would be easy to figure out the barcode and UMI length without further help, given that the two sequences are (i) already present in the read name and (ii) separated by an underscore. This would allow users to specify the barcode structure once, which improves usability and simplifies workflows.

**Comment by Shians** [comment]
True, it should be detectable from the sequences. I've written a preliminary implementation for guessing the barcode and UMI lengths based on the first read.

```
std::pair<int, int> get_bc_umi_lengths(string bam_fn) {
    BGZF *fp = bgzf_open(bam_fn.c_str(), "r"); // input file
    bam_hdr_t *bam_hdr = bam_hdr_read(fp);

    bam1_t *bam_record = bam_init1();

    if (bam_read1(fp, bam_record) >= 0) {
        string read_header = bam_get_qname(bam_record);
        int break_pos = read_header.find("#");
        // start from 1 to exclude @
        string first_section = read_header.substr(1, break_pos);
        bool valid_pattern = std::regex_search(first_section, std::regex("[ACTGN]+_[ACRGN]+"));
        if (!valid_pattern) {
            throw std::runtime_error("Read header does not contain valid barcode-umi data.");
        }

        // header has structure {BARCODE}_{UMI}
        int bc_len = first_section.find("_") + 1;
        int umi_len = first_section.length() - bc_len - 1;

        Rcout << "detected barcode length: " << bc_len << "\n";
        Rcout << "detected UMI length: " << umi_len << "\n";

        return std::make_pair(bc_len, umi_len);
    }
    else
    {
        throw std::runtime_error("BAM file reading failed.");
    }
}
```

The returned pair can be assigned using

```
int bc_len;
int UMI_len;
std::tie(bc_len, UMI_len) = get_bc_umi_lengths(bam_fn);
```

Because this leads to a change in the API I'm inclined to leave it in the shadows until we do some kind of 2.0.0 release. Also slightly worried that guessing from just the first read is not very robust, but I'm pretty sure we segfault anyway when any reads are malformed.

**Comment by LTLA** [comment]
Well, it's worse than segfaulting. I was getting "UMIs" that contained an underscore, which was clearly nonsensical; but the pipeline kept on going silently and gave me an empty count matrix without any errors.

**Comment by LuyiTian** [comment]
The separate functions are designed so scPipe can process bam file generated from Drop-seq tools and CellRanger, which also uses bam file and aux tag to store barcode and UMI. And the original plan is to write an integrated function so for most time users don't need to touch individual functions (https://github.com/LuyiTian/scPipe/commit/f79d6e6e7b584351a8a6a89a562d432f148fae4c).  However, the integrated function will require Rsubread, which is not available on Win, then the scPipe RCheck would fail on Win. So I had to remove that function so scPipe can be accepted by Bioconductor. 

It would be great if you have an idea of adding alignment step without breaking R check.

**Comment by LTLA** [comment]
I don't think the problem is with the lack of an integrated function. I'm happy to call each step separately. The problem is that `sc_exon_mapping` is not correctly re-using the already specified information.

The simplest solution is to simply not have defaults for the barcode and UMI lengths in `sc_exon_mapping`. This means that users are forced to re-specify them, which is annoying but better than a silent error.

The better solution would be to determine the lengths from the read names when it is clear that the read names have been generated by `sc_trim_barcode`. If the names are not of a valid format, the function can fall back to something else, e.g., extracting the barcodes/UMIs from BAM tags or whatever.
